lib/ui/semantics.dart

```diff
@@ -641,6 +641,169 @@ class SemanticsFlag {
+       ],
+       assert(
+         range == null ||
+         (range.start <= string.length && range.end <= string.length),
+         'The range $range is outside of the string $string'
+       );
+
+  /// The plain string.
+  final String string;
+
+  /// The attributes the [string] carries.
+  final List<StringAttribute>? attributes;
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    return other is AttributedString
+        && other.string == string
+        && other.attributes == attributes;
+  }
+
+  @override
+  int get hashCode => hashValues(string, attributes);
+
+  @override
+  String toString() {
+    return '$string${attributes != null && attributes!.isNotEmpty ? ' (attributes: $attributes)': ''}';
+  }
+}
+
+
+// When adding a new StringAttributeType, the classes in these file must be
+// updated as well.
+//  * engine/src/flutter/lib/web_ui/lib/src/ui/semantics.dart
+//  * engine/src/flutter/lib/ui/semantics/attributed_string.h
+//  * engine/src/flutter/shell/platform/android/io/flutter/view/AccessibilityBridge.java
+
+/// The enum that is used in [StringAttribute] class.
+enum StringAttributeType {
+  /// The string should be spell out character by character.
+  ///
+  /// The additional argument of the string attribute is null.
+  spellOut,
+  /// The string should be treated as a specific language.
+  ///
+  /// The additional argument of the string attribute is a string language tag.
+  locale,
+}
+
+/// An abstract interface for string attribute that affects how assistive
+/// technologies, e.g. VoiceOver or TalkBack, treat the text.
+class StringAttribute {
+  StringAttribute._({
+    required this.range,
+    required this.type,
+    required this.args,
+  });
+
+  /// A contstructor to create a string attribute that causes the assistive
+  /// technologies, e.g. VoiceOver, to spell out the string character by character.
+  StringAttribute.spellOut({
+    required this.range
+  }) : type = StringAttributeType.spellOut,
+       args = null,
+       assert(range.isValid);
+
+  /// A contstructor to create a string attribute that causes the assistive
+  /// technologies, e.g. VoiceOver, to treat string as a certain language.
+  StringAttribute.locale(
+    Locale locale, {
+    required this.range
+  }) : type = StringAttributeType.locale,
+       args = locale.toLanguageTag(),
+       assert(range.isValid);
+
+  /// The text range to which this attribute should apply.
+  final TextRange range;
+
+  /// The enum that indicates what this attribute type is.
+  final StringAttributeType type;
+
+  /// The additional arguement this attribute carries.
+  ///
+  /// Depends on what [type] this attribute is, it may or may not carry an
+  /// additional arguement. For which more information, see the [StringAttributeType].
+  final Object? args;
+
+  /// Create a new copy of this string attribute with the given range.
+  StringAttribute copyWithRange(TextRange range) {
```

Comment:
nit: could just be named "copy" with an optional range argument. If non is provided, the current range is used. I think that's the copy pattern we sue elsewhere.